### PR TITLE
Protect login from brute force

### PR DESCRIPTION
**Pull request by hedgehog1833**
Failed logins are cached for 24 hours (hashed IP and the amount of attemps). If there are more than five failed attempts the IP is blocked for 24 hours.

### CODE REVIEW

**Comment by hedgehog1833** [comment]
[[Detector] Captcha after too many failed logins](https://trello.com/c/9NdqWhND/153-detector-captcha-after-too-many-failed-logins)

File: webapp/src/test/java/rocks/metaldetector/security/LoginAttemptServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import org.springframework.cache.CacheManager;
+
+import static org.mockito.Mockito.reset;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static rocks.metaldetector.security.LoginAttemptService.FAILED_LOGINS_CACHE;
+
+@ExtendWith(MockitoExtension.class)
+class LoginAttemptServiceTest implements WithAssertions {
```
Comment (DanielW1987) [inline webapp/src/test/java/rocks/metaldetector/security/LoginAttemptServiceTest.java:22]: Please have a look at the namings in this class. In my opinion, no user or user account is blocked. It's only the IP.

File: webapp/src/test/java/rocks/metaldetector/service/user/UserServiceTest.java [hunk 2]
```diff
@@ -345,6 +358,21 @@ void get_user_by_email_or_username_for_not_existing_user() {
       assertThat(userDto).isEmpty();
     }
 
+    @Test
+    @DisplayName("Requesting a blocked user by email or username should throw an exception")
+    void get_user_by_email_or_username_for_blocked_user() {
```
Comment (DanielW1987) [inline webapp/src/test/java/rocks/metaldetector/service/user/UserServiceTest.java:363]:
```suggestion
    @DisplayName("Requesting a user by email or username via blocked IP should throw an exception")
    void get_user_by_email_or_username_from_blocked_ip() {
```

File: webapp/src/main/java/rocks/metaldetector/security/LoginAttemptService.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  public boolean isBlocked(String ipHash) {
+    Cache cache = cacheManager.getCache(FAILED_LOGINS_CACHE);
+
+    if (cache != null) {
+      Cache.ValueWrapper valueWrapper = cache.get(ipHash);
+      return valueWrapper != null && (int) valueWrapper.get() >= MAX_ATTEMPTS;
+    }
+    return true;
```
[the remark this answers is not in the document]
Comment (hedgehog1833) [reply webapp/src/main/java/rocks/metaldetector/security/LoginAttemptService.java:39]: I would say this is a `should not happen` anyway and from a security point of view I would rather block too much than not enough.

**Pull request merged**